Handle shadowing of bindings through macroexpansion

czan opened the pull request:
I've added two test cases to show the type of expansion that breaks without this change. The crux of it is that we need to provide an environment to `macroexpand`, but it doesn't let us do so. To get it to work we cheat a bit and invoke the macro function itself in [approximately the same way as Compiler.java](https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/Compiler.java#L6992-L6993) does.

Commit:
Handle shadowing of bindings through macroexpansion

mszajna commented:
Big thank you for your contribution! I'm reviewing your changes right now and I'll try to push out an updated version as soon as I can.

I would like to make sure I understand the underlying issue though and see if there's anything about the generative testing I could improve not to miss such cases in the future.

mszajna commented on src/await_cps/ioc.clj:
```diff
@@ -39,20 +39,30 @@
+  ;; This adds mappings to "true" into the environment map. This doesn't quite
+  ;; match what the Clojure compiler does, but I think it's already recommended
+  ;; that macros don't depend on the values in the &env map.
+  (update ctx :env (fnil into {}) (map (fn [sym] [sym true])) syms))
```
What does `(resolve env sym)` ([line 4](https://github.com/mszajna/await-cps/pull/1/files#diff-4944141e3afac9160873219948a5bebcR4)) do with symbols mapped to `true`?

mszajna replied:
It appears to be resolving to `nil` but I can see now it doesn't matter as you only care for it not to resolve to certain vars.

czan commented:
The issue is when var resolution (with `resolve`) doesn't take into account locally established bindings. This isn't a problem for run-time values, because the rewritten code will maintain the visible bindings and everything will work out. Clojure allows macros to be shadowed, though, so that means that during macroexpansion the local bindings (ie. `&env`) need to be taken into account, too. Due to the way `async` recursively macroexpands we need to augment `&env` to also include the other binding forms that we see along the way.

In terms of adding it to the code generation, I think it would suffice to have a macro defined, and then generate calls to the macro and bindings of the same name. Something like:

```clojure
;; define the macro at the top level
(defmacro test-macro [x] `(inc ~x))

;; then teach the generators to generate forms like:
`(~'test-macro ~a-number) ;; we have to be careful with the namespace of the symbol
`(let [test-macro (fn [x#] (- x#))] ~another-generated-form)
```

mszajna commented on test/await_cps_test.clj:
```diff
@@ -205,6 +205,14 @@
+(deftest shadowing-symbols
+  ;; Here we shadow the `await` macro with a local function
+  (is (= 1 (:value (run-async timeout `(let [~'await (constantly 1)]
+                                         (~'await (fn [r# e#] (r# 2))))))))
```
The assertion above seems to pass both with and without your patch applied if the namespace you're running the test from does not map `await` to `await-cps/await`. That's the case when loading a fresh REPL and kicking off tests from `user` namespace unless you `user=> (require '[await-cps :refer [await]])`.

Any ideas how to make it reproducible regardless of environment?

czan replied:
Yeah, this isn't too hard. If we bind `*ns*` in the test it will resolve properly. I'll push a commit to fix this soon.

mszajna commented:
Thanks for clarifying. I'm going to give code gen a longer think - it's likely to be somewhat tricky with the interplay of namespace mappings and macroexpansion as per the comment on `shadowing-symbols` test.

Another 2 binding constructs, `loop` and `catch`, are likely affected by shadowing issues as well. Would you be happy to cover them too or rather have me pick it up from there?

I'm otherwise satisfied with this PR and unless you want to update it further I'm inclined to merge it in. Thanks again for your contribution and thanks for engaging with this library in the first place! I would love to hear about your use-case, if you don't mind of course.

czan commented:
I'd like to fix up `loop` and `catch` before merging. I think the namespace mapping issue can be solved by binding `*ns*` in the test (like I suggested for the `shadowing` one above), so I would also love to get the code gen part in if we can. I'll see if I can get something working.

Commit:
Fix shadowing in other binding forms (loop*, catch)

Commit:
Fix the namespace used in the shadowing-symbols test

czan commented:
Okay, I've added support for `loop*` and `catch`, and fixed the test, but I took a quick look at the generators and it's going to be more involved than I thought, so I'm not going to do that right now. I'm happy for you to merge this PR whenever you want.
